File: stronghold/code/python/DNA.py
import os



class DNASolution:
    def __init__(self):
        self.dna = None
        self.dna_bases = {'A': 0, 'C': 0, 'G': 0, 'T': 0}

        self.read_dna_seq()
        self.count_dna_bases()

    # lendo a sequência do dna
    def read_dna_seq(self):
        try:
            path = os.path.realpath("../../input/rosalind_dna.txt")
            input_data = open(path, 'r')
            self.dna = input_data.read().strip()
            input_data.close()
        except IOError:
            print("Erro na leitura do arquivo!")

    # contando a qtd. de cada base (A, C, G, T)
    def count_dna_bases(self):
        for base in self.dna:
            self.dna_bases[base] += 1

    # resultado da qtd. de cada base
    def number_dna_bases(self):
        bases = list(self.dna_bases.values())

        try:
            path = os.path.realpath("../../output/DNA.txt")
            output_data = open(path, 'w')
            output_data.write(' '.join(map(str, bases)))
            output_data.close()
        except IOError:
            print("Erro na escrita do arquivo!")


if __name__ == "__main__":
    solution = DNASolution()
    solution.number_dna_bases()

File: stronghold/code/dart/DNA.dart
import 'dart:io';

class DNASolution {
  String currentDir = Directory.current.path;

  late String dna;
  Map<String, int> dnaBases = {'A': 0, 'C': 0, 'G': 0, 'T': 0};

  DNASolution() {
    initialize_dna_sequence();
  }

  // 'pegando' a sequência
  Future<void> initialize_dna_sequence() async {
    try {
      dna = await get_dna_sequence();
      count_dna_bases();
    } catch (e) {
      print("Erro na leitura do arquivo! $e");
    }
  }

  // sequência de DNA
  Future<String> get_dna_sequence() async {
    // caminho do arquivo
    String path = "$currentDir/stronghold/input/rosalind_dna.txt";
    // arquivo
    File inputData = File(path);

    String dnaSequence = await inputData.readAsString();
    return dnaSequence.trim();
  }

  // contando a quantidade de cada base da sequência
  void count_dna_bases() {
    for (var i = 0; i < dna.length; i++) {
      // base
      String base = dna[i];

      // realizando a contagem
      dnaBases.update(base, (value) => value + 1);
    }

    number_dna_bases();
  }

  // número de cada base
  void number_dna_bases() async {
    var numbers = await dnaBases.values;

    try {
      // caminho do arquivo
      String path = "$currentDir/stronghold/output/DNA.txt";
      // arquivo
      File outputData = File(path);
      var f = outputData.openWrite();
      f.write(numbers.join(' '));
      await f.close();
    } catch (e) {
      print("Erro na escrita do arquivo! $e");
    }
  }
}

void main() {
  DNASolution();
}
